tools/workflow/cleaner/apigw/clean.go

```diff
@@ -0,0 +1,80 @@
+package apigw
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/apigateway"
+	"log"
+	"os"
+	"regexp"
+	"time"
+)
+
+const Type = "apigw"
+
+var logger = log.New(os.Stdout, fmt.Sprintf("[%s] ", Type), log.LstdFlags)
+
+func Clean(sess *session.Session, expirationDate time.Time) error {
+	logger.Printf("Begin to clean APIGW clients")
+
+	apigwclient := apigateway.New(sess)
+	var deleteGatewayIds []*string
+	var nextToken *string
+	for {
+		getRestApisInput := &apigateway.GetRestApisInput{Position: nextToken}
+		getRestApisOutput, err := apigwclient.GetRestApis(getRestApisInput)
+		if err != nil {
+			return fmt.Errorf("unable to retrieve APIs: %w", err)
+		}
+		for _, gw := range getRestApisOutput.Items {
+			doesNameMatch, err := shouldDelete(gw)
+			if err != nil {
+				return fmt.Errorf("error during pattern match: %w", err)
+			}
+			if expirationDate.After(*gw.CreatedDate) && doesNameMatch {
+				logger.Printf("Try to delete gateway %s created-at %v", *gw.Id, gw.CreatedDate)
+				deleteGatewayIds = append(deleteGatewayIds, gw.Id)
+			}
+		}
+		if getRestApisOutput.Position == nil {
+			break
+		}
+		nextToken = getRestApisOutput.Position
+
+		if len(deleteGatewayIds) < 1 {
+			logger.Printf("No API Gateways to delete")
+			return nil
+		}
+
+		for _, id := range deleteGatewayIds {
+			terminateGatewayInput := &apigateway.DeleteRestApiInput{RestApiId: id}
+			if _, err := apigwclient.DeleteRestApi(terminateGatewayInput); err != nil {
+				return fmt.Errorf("unable to delete workspace: %w", err)
+			}
+			// 30 second request limit to delete ap
+			// https://docs.aws.amazon.com/apigateway/latest/developerguide/limits.html#api-gateway-control-service-limits-table
+			time.Sleep(31 * time.Second)
```

Comment:
I'm just curios if deleting while you still paginating over the existing API Gateways will not causes any problem in the pagination.

I think it is safer to paginate and collect the full list and then delete?